chain/indexer/tipset.go

```diff
@@ -0,0 +1,408 @@
+	"github.com/filecoin-project/lily/tasks/messages/message"
+	"github.com/filecoin-project/lily/tasks/messages/parsed_message"
+	"github.com/filecoin-project/lily/tasks/messages/receipt"
+	"github.com/filecoin-project/lily/tasks/msapprovals"
+)
+
+var tsLog = logging.Logger("lily/index/tipset")
+
+// TipSetIndexer waits for tipsets and persists their block data into a database.
+type TipSetIndexer struct {
+	name  string
+	node  tasks.DataSource
+	tasks []string
+
+	procBuilder *processor.Builder
+}
+
+type TipSetIndexerOpt func(t *TipSetIndexer)
+
+// NewTipSetIndexer extracts block, message and actor state data from a tipset and persists it to storage. Extraction
+// and persistence are concurrent. Extraction of the a tipset can proceed while data from the previous extraction is
+// being persisted. The indexer may be given a time window in which to complete data extraction. The name of the
+// indexer is used as the reporter in the visor_processing_reports table.
+func NewTipSetIndexer(node tasks.DataSource, name string, tasks []string, options ...TipSetIndexerOpt) (*TipSetIndexer, error) {
+	tsi := &TipSetIndexer{
+		name:  name,
+		node:  node,
+		tasks: tasks,
+	}
+
+	tipsetProcessors := map[string]processor.TipSetProcessor{}
+	tipsetsProcessors := map[string]processor.TipSetsProcessor{}
+	actorProcessors := map[string]processor.ActorProcessor{}
+	reportProcessors := map[string]processor.ReportProcessor{
+		"builtin": indexer.NewTask(node),
+	}
+
+	for _, t := range tasks {
+		switch t {
+		//
+		// miners
+		//
+		case MinerCurrentDeadlineInfoTask:
+			actorProcessors[t] = actorstate.NewTask(node, actorstate.NewTypedActorExtractorMap(
+				miner.AllCodes(), miner2.DeadlineInfoExtractor{},
+			))
+		case MinerFeeDebtTask:
+			actorProcessors[t] = actorstate.NewTask(node, actorstate.NewTypedActorExtractorMap(
+				miner.AllCodes(), miner2.FeeDebtExtractor{},
+			))
+		case MinerInfoTask:
+			actorProcessors[t] = actorstate.NewTask(node, actorstate.NewTypedActorExtractorMap(
+				miner.AllCodes(), miner2.InfoExtractor{},
+			))
+		case MinerLockedFundsTask:
+			actorProcessors[t] = actorstate.NewTask(node, actorstate.NewTypedActorExtractorMap(
+				miner.AllCodes(), miner2.InfoExtractor{},
+			))
+		case MinerPreCommitInfoTask:
+			actorProcessors[t] = actorstate.NewTask(node, actorstate.NewTypedActorExtractorMap(
+				miner.AllCodes(), miner2.PreCommitInfoExtractor{},
+			))
+		case MinerSectorDealTask:
+			actorProcessors[t] = actorstate.NewTask(node, actorstate.NewTypedActorExtractorMap(
+				miner.AllCodes(), miner2.SectorDealsExtractor{},
+			))
+		case MinerSectorEventsTask:
+			actorProcessors[t] = actorstate.NewTask(node, actorstate.NewTypedActorExtractorMap(
+				miner.AllCodes(), miner2.SectorEventsExtractor{},
+			))
+		case MinerSectorPoStTask:
+			actorProcessors[t] = actorstate.NewTask(node, actorstate.NewTypedActorExtractorMap(
+				miner.AllCodes(), miner2.PoStExtractor{},
+			))
+		case MinerSectorInfoTask:
```

Comment:
@iand Here is where I've combined tables produced by the `miner_sector_info` task arg, passing this task name will produce models for both the `miner_sector_info` and `miner_sector_info_7v` table.